DOCSUP-15237: Documentation of the exponentialMovingAverage function (#31396)
ClickHouse/ClickHouse: docs/en/sql-reference/aggregate-functions/reference/exponentialmovingaverage.md

---
toc_priority: 108
---

## exponentialMovingAverage {#exponential-moving-average}

> An aggregate function that calculates the exponential moving average of values for the determined time. 

Reviewer: ```suggestion
Сalculates the exponential moving average of values for the determined time. 
```

**Syntax**

```sql
exponentialMovingAverage(x)(value, timestamp)
```

Each `value` corresponds to the determinate `timestamp`. The half-life `x` is the time lag at which the exponential weights decay by one-half. The function returns a weighted average: the older the time point, the less weight the corresponding value is considered to be.

**Arguments**

-   `value` — Value. [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md) or [Decimal](../../../sql-reference/data-types/decimal.md).

> -   `timestamp` — Timestamp must. [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md) or [Decimal](../../../sql-reference/data-types/decimal.md).

Reviewer: ```suggestion
-   `timestamp` — Timestamp. [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md) or [Decimal](../../../sql-reference/data-types/decimal.md).
```

**Parameters**

-   `x` — Half-life period. [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md) or [Decimal](../../../sql-reference/data-types/decimal.md).

**Returned values**

> - Returnes an exponentially smoothed moving average of the values for the past `x` time at the latest point of time (https://en.wikipedia.org/wiki/Unix_time).

Reviewer: ```suggestion
- Returnes an [exponentially smoothed moving average](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average) of the values for the past `x` time at the latest point of time (https://en.wikipedia.org/wiki/Unix_time).
```

Type: [Float64](../../../sql-reference/data-types/float.md#float32-float64).

> **Example**

Reviewer: ```suggestion
**Examples**
```

Input table:

``` text
┌──temperature─┬─timestamp──┐
│          95  │         1  │
│          95  │         2  │
│          95  │         3  │
│          96  │         4  │
│          96  │         5  │
│          96  │         6  │
│          96  │         7  │
│          97  │         8  │
│          97  │         9  │
│          97  │        10  │
│          97  │        11  │
│          98  │        12  │
│          98  │        13  │
│          98  │        14  │
│          98  │        15  │
│          99  │        16  │
│          99  │        17  │
│          99  │        18  │
│         100  │        19  │
│         100  │        20  │
└──────────────┴────────────┘
```

Query: 

```sql

> exponentialMovingAverage(5)(temperature, timestamp)

Reviewer: ```suggestion
SELECT exponentialMovingAverage(5)(temperature, timestamp);
```

```

Result:

``` text
┌──exponentialMovingAverage(5)(temperature, timestamp)──┐
│                                    92.25779635374204  │
└───────────────────────────────────────────────────────┘
```

**Example 2**

> 

Reviewer: ```suggestion
```

Query: 

```sql

> SELECT

Reviewer: ```suggestion
SELECT  value, time, round(exp_smooth, 3), bar(exp_smooth, 0, 1, 50) AS bar FROM (SELECT (number = 0) OR (number >= 25) AS value, number AS time, exponentialMovingAverage(10)(value, time) OVER (Rows BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS exp_smooth FROM numbers(50));
```

---
toc_priority: 108
---

## exponentialMovingAverage {#exponential-moving-average}

An aggregate function that calculates the exponential moving average of values for the determined time. 

**Syntax**

```sql
exponentialMovingAverage(x)(value, timestamp)
```

> Each `value` corresponds to the determinate `timestamp`. The half-decay period is the time interval `x` during which the previous values are taken into account. The function returns a weighted average: the older the time point, the less weight the corresponding value is considered to be.

Reviewer 2: I think `half-life` is more clear than `half-decay`

Author: Fixed
![image](https://user-images.githubusercontent.com/92211164/142471746-205f10ca-bcca-4bab-81a6-5654953eac13.png)

> **Arguments**

Reviewer: ```suggestion
**Arguments**

```

> - `value` - value must be [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md), or [Decimal](../../../sql-reference/data-types/decimal.md).

Reviewer: ```suggestion
- `value` — Value must be [Integer](../../../sql-reference/data-types/int-uint.md), [Float](../../../sql-reference/data-types/float.md), or [Decimal](../../../sql-reference/data-types/decimal.md).
```

> - `timestamp` - timestamp must be [Integer](../../../sql-reference/data-types/int-uint.md).

Reviewer: ```suggestion
- `timestamp` — Timestamp must be [Integer](../../../sql-reference/data-types/int-uint.md).
```

**Parameters**

> - `x` - half-decay period in seconds must be  [Integer](../../../sql-reference/data-types/int-uint.md).

Reviewer: ```suggestion

-   `x` — Half-life period in seconds must be [Integer](../../../sql-reference/data-types/int-uint.md).
```

**Returned value(s)**

> - Returnes an exponentially smoothed moving average of the values for the past `x` time at the latest point of time.

Reviewer 2: https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average

Author: Added
